x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4HttpServerTransportCloseNotifyTests.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.transport.netty4;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpObjectAggregator;
+import io.netty.handler.codec.http.HttpRequestEncoder;
+import io.netty.handler.codec.http.HttpResponseDecoder;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.util.concurrent.Future;
+
+import org.elasticsearch.common.network.NetworkService;
+import org.elasticsearch.common.settings.MockSecureSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.env.TestEnvironment;
+import org.elasticsearch.http.AbstractHttpServerTransportTestCase;
+import org.elasticsearch.http.HttpServerTransport;
+import org.elasticsearch.http.netty4.Netty4HttpServerTransport;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.telemetry.tracing.Tracer;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.netty4.SharedGroupFactory;
+import org.elasticsearch.transport.netty4.TLSConfig;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.net.ssl.SSLException;
+
+public class SecurityNetty4HttpServerTransportCloseNotifyTests extends AbstractHttpServerTransportTestCase {
```

Comment:
Tests make sense, but I think we'll also need some integ tests showing how this interacts with the rest of the system, particularly the task cancellation mechanism (which it doesn't today, but it'll need to).

Reply:
I'm still thinking should we cancel or not. If we serve all requests while channel is writable then we dont need additional handling for cancelation. If channel is not writable, client decided not to wait, it will go through exception path, which exists today.

Reply:
Although it is certainly possible to keep serving responses as you propose, it's also legitimate to cancel tasks and send the resulting exception responses, or indeed to just tear down the connection without sending any of the remaining responses, so this is more of a design question than a technical absolute.

I couldn't find any clients which might send a `close_notify` while awaiting responses - AFAICT this only happens on the abort path or when dropping an idle connection from a pool.

Reply:
You are right, looking further into HTTP and TLS RFC's closing connection on `close_notify` make sense. Serving all responses and closing connection should be handled with `Connection: close` header in HTTP1. In TLS prior to v1.3 `close_notify` should discard all pending writes and close connection. TLS v1.3 tries to address [truncation attack](https://en.wikipedia.org/wiki/Transport_Layer_Security#:~:text=142%5D%5B143%5D-,Truncation%20attack,-%5Bedit%5D) by letting application layer to figure out how to close connection,  I'm not sure how it's applicable to our server.

Reply:
I don't think truncation attacks are a possibility here, we always send either a `Content-length` header or else a properly-terminated chunked-encoded message so clients can tell if a response is incomplete at the HTTP layer. But this isn't the right place to discuss this more deeply - if you think there's a security issue in this area then please [mail security@elastic.co](https://www.elastic.co/community/security) with details.

Reply:
Changed code to close connection immediately in all cases. Found that `SslHandler` triggers `userEvent` on `close_notify` `SslCloseCompletionEvent`, so I updated `Netty4HttpPipeliningHandler` to listen this event and close connection.

Reply:
Also added integ test - SecurityNetty4TransportCloseNotifyIT. It's pretty similar to unit test, but on internalCluster. Have to use custom http client to send close_notify, I tried to hijack apache's http client sslEngine, but it got pretty messy.

>  I think we'll also need some integ tests showing how this interacts with the rest of the system, particularly the task cancellation mechanism (which it doesn't today, but it'll need to).

Since there is no cancellation now, is there something I should do with current tests?

Reply:
> Since there is no cancellation now,

I think task cancellation should be hooked-up by default with this implementation. Since you close the channel, that will propagate through the listener framework and cancel tasks (I think).

`Netty4HttpChannel#addCloseListener`.

Reply:
Nice! Is it worth to test cancellation in integ test? If yes, what rest operation would be a good example?

Reply:
> Nice! Is it worth to test cancellation in integ test?

I will take another look at this PR later today. 

But IMO, if "HTTP Close triggers Task cancellation" triggers task cancellation is already tested, then this PR only needs to validate the "SSL client close_notify" leads to channel close.

However, I don't know if that test exists. My initial perusal of the code is that we do not have this test at an HTTP level (which we should). I would say this PR should implement it if it is missing.

Reply:
I would look in `RestActionCancellationIT` and other classes in that qa package (`SearchRestCancellationIT`,  `ClusterHealthRestCancellationIT`, etc).

It looks like these are testing task cancellation through the http layer. I would do a bit of investigation and look at how the "cancellation" is implemented. I _think_ it might be through closing the http connection. But that appears to be a bit deep in the apache code and we should confirm it.

If those tests are sufficient then I think we are covered at the moment.

Reply:
I added action cancellation test.

`RestActionCancellationIT` and `SearchRestCancellationIT` call request.abort() which indeed close connection. 
With help of neat plugin `CancellableActionTestPlugin` I found it's pretty easy to add integ test here for action cancellation - `testSendCloseNotifyCancelAction`